man/man1/xtfs_scrub.1
.TH xtfs_scrub 1 "July 2011" "The XtreemFS Distributed File System" "XtreemFS Admin Tools"
.SH NAME
xtfs_scrub \- checks for each file in a volume if the MRC file size is outdated and if the checksum of all objects is correct. The scrubber can also replace defective replicas in repair mode. 
.SH SYNOPSIS
\fBxtfs_scrub [ \fIoptions\fB ] \fI<volume_name>\fB
.br

.SH DESCRIPTION
.I xtfs_scrub
performs a consistency check of each file in a given volume. If run without \-repair or \-delete it will only perform checks and report problems. Th scrubber checks for file if the file size on the OSDs is stored on the MRC and fixes inconsistencies if run in the \-repair mode. In addition, the scrubber checks the checksum for each object of a file if the checksums are enabled on the OSD.

For replicated files the scrubber will check each replica for objects with inccorect checksum. In addition, it marks full replicas that successfully copied all objects as "complete" replicas. In the \-repair mode, the scrubber will also replace replicas that contain ramoved OSDs. An OSD must be marked as dead/removed with the xtfs_chstatus tool, otherwise the scrubber will consider the OSD as temporarily unavailable and won't replace the replica.

.SH EXAMPLE USAGE
.B "xtfs_scrub \-dir pbrpc://localhost:32638 myVolume
.PP
Scrubs all files in the volume named 'myVolume' registered at the Directory Service 'pbrpc://localhost:32638'.

.SH OPTIONS
.TP
.TP
\fB-c \fI<creds_file>
Path to a PKCS#12 credentials file (private key + certificate) to use for SSL authentication. Must be present when MRC URL starts with pbrpcs:// or pbrpcg://.
.TP
\fB-cpass \fI<creds_passphrase>
An optional passphrase to access the credentials file.
.TP
\fB\-delete
Deletes files where the objects are lost because they were stored on a failed/removed OSD.
.TP
\fB-dirs \fI<uri(,uri...)>
Specifies the directory services to use (e.g. 'pbrpc://localhost:32638,pbrpc://myDir.com:32638') (comma-separated).
If no URI is specified, URIs and security settings are taken from '/etc/xos/xtreemfs/default_dir'.
In case of a secured URI ('pbrpcs://...' or 'pbrpcg://...'), it is necessary to also specify SSL settings.
.TP
\fB-h/--help
Shows usage info.
.TP
\fB\-repair
Repairs inconsistent file sizes and replace replicas on failed/removed OSD.
.TP
\fB\-silent
Hides detailed information about the scrubbed files.
.TP
\fB-t \fI<trusted_CAs>
Path to a PKCS#12 file containing a set of certificates from trusted certification authorities. These certificates will be used to authenticate the MRC. Must be present when MRC URL starts with pbrpcs:// or pbrpcg://.
.TP
\fB\-thrs \fI<#threads> 
Sets the number of threads to use for scrubbing (default=10).
.TP
\fB-tpass \fI<trusted_passphrase>
An optional passphrase to access the truststore file.
.TP
\fB---ssl-protocol \fI<ssl-protocol> 
SSL/TLS version to use: ssltls, tlsv1, tlsv11, tlsv12. 'ssltls' (default) accepts all versions, the others accept only the exact version they name. 'tlsv12' is available in JDK 7+ only. 'tlsv11' comes with JDK 6 or 7, depending on the vendor.

.SH "SEE ALSO"
.BR xtfs_chstatus (1),
.BR xtfs_cleanup (1),
.BR xtfs_mrcdbtool (1),
.BR xtfs_remove_osd (1),
.BR

.SH AVAILABILITY
The xtfs_scrub command is part of the XtreemFS-tools package and is available from \fIhttp://www.xtreemfs.org\fP.


CHANGELOG
26-OCT-2016: release 1.6.0 (Phantastic Pancakes)

  Changes:
    * Extended volume quota implementation to OSDs. This requires dumping and reimporting the MRC database during upgrade (see 'MRC Database Backups and Conversion' in the user guide).
      * Implemented hard quotas for volumes and user and groups per volume.
      * Added quota information to import and export function.
      * These changes supersede the previous implementation.
    * Fix for mount.xtreemfs to fully support mount options via /etc/fstab.
    * Introduced Java wrappers for the C++ version of libxtreemfs and made it the default Java implementation.
    * Improved Hadoop adapter to work with a broader range of systems and versions.
    * Added 'listen.port.bind_retries' server option that implements exponential backoff during service startup if the port is not free yet.
    * Added RabbitMQ OSD tracing policy (ID 6003).
    * Added Preferred UUID OSD selection policy (ID 1003).
    * Extended X509 authentication to allow user and group mapping.
    * Slimmed down and updated Mac OS X installer with stricter dependency checking and SIP awareness.
    * Converted the Java parts to Maven and made all components available as artifacts hosted on GitHub.
    * Cleanup of code base to exclude generated sources and previously shipped binaries.
    * Add non-interactive console support to command line tools.
    * Remove explicit support for SSLv3.
    * Many small stability improvements and fixes.

12-MAR-2015: release 1.5.1

  Changes:
    * Hadoop Adapter supports Hadoop-2.x and other applications running on the YARN platform (tested with Apache Spark and Apache Flink), new OSD selection policy that imitates the HDFS placement behavior.
    * Consistent adding and removing replicas for R/W replication.
    * xtfs_scrup now replaces failed replicas using the R/W replication policy.
    * Improvements of the SSL mode: the used SSL/TLS version is selectable, strict certificate chain checks are possible, the SSL code on client and server side was improved.
    * Support for PKCS12 certificate containers is remove from the Windows client (did not work in previous release).
    * mount.xtreemfs accepts all parameters to be passed in the form -o option=value. If XtreemFS is mounted via /etc/fstab, mount options are passed to the XtreemFS client.
    * Dockerfiles for the XtreemFS services are available at https://github.com/xtreemfs/xtreemfs-docker.
    * A Vagrantfile is available in the XtreemFS repository to ease development and testing for new contributors.
    * Fixed snapshots in the case of deleted files.
    * Snapshots can be deleted even after inactivating snapshots for a volume (issue 290).
    * Initial version of an LD_PRELOAD based client that bypasses FUSE (experimental).
    * Initial implementation of volume quotas (checked only while opening files by the MRC).
    * OSDs can report their health to the DIR (e.g. based on SMART values), which aggregates this information in the web interface. A new OSD selection policy sorts out unhealthy OSDs.
    * Improved automatic testing: introduced Travis-CI, migrated all JUnit tests to version 4, improved reliability of nightly integration tests.
    * Added new benchmark tool xtfs_benchmark that is based on the JAVA libxtreemfs.
    * Many small improvements and fixes.

12-MAR-2014: release 1.5 (Wonderful Waffles)

  Changes:
    * Improved Hadoop Adapter e.g., added support for multiple volumes and read and write buffer to speed up performance.
    * Support for SSDs: Support for multiple OSD storage threads to increase parallelism.
    * Status webpage as part of the DIR webinterface for replicated files to show current primary and backup replicas.
    * Multi-homing support. XtreemFS can now be made available for multiple networks and clients will pick the correct address automatically.
    * Support for multiple OSDs per machine (e.g. one for each disk) through the xtreemfs-osd-farm init.d script.
    * Fixed major issues in Read-Only and Read/Write file replication.
    * Reduced metadata cache default timeout from 120 to 10 seconds. Disable it completely with --metadata-cache-size=0.
    * Upgraded to Google Protobuffers 2.5 which improves performance.
    * Windows Client: Fixed several issues (e.g., 286, 299) and updated the CbFS driver to the latest version.
    * XtreemFS also compiles under buildroot now.
    * Fixed several issues around SSL (e.g., 263, 281).
    * Fixed many small issues e.g., reported volume capacity was inaccurate (issue 296).
      Other fixed issues: 267, 268, 272, 274, 276, 277, 278, 283, 287, 288, 292, 293, 294.
    * Improved general stability (e.g., see issues 297, 301).
    * xtfs_scrub now repairs replicas with an invalid checksum automatically.
    * Disabled deferred deletion of file data on the OSDs.
    * Improved test coverage of nightly tests and added missing unit tests and a Valgrind leak check for C++ unit tests.

12-NOV-2012: release 1.4 (Salty Sticks)

  Changes:
    * Improved stability: Fixed client crashes when timeouts occurred; fixed server crashes and issues with the R/W replication.

java/servers/src/org/xtreemfs/utils/xtfs_scrub/xtfs_scrub.java
/*
 * Copyright (c) 2009-2009 by Bjoern Kolbeck, Nele Andersen,
 *                            Zuse Institute Berlin
 *
 * Licensed under the BSD License, see LICENSE file for details.
 *
 */

package org.xtreemfs.utils.xtfs_scrub;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Collection;
import java.util.EmptyStackException;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.Stack;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;

import org.xtreemfs.common.libxtreemfs.AdminClient;
import org.xtreemfs.common.libxtreemfs.AdminVolume;
import org.xtreemfs.common.libxtreemfs.ClientFactory;
import org.xtreemfs.common.libxtreemfs.ClientFactory.ClientType;
import org.xtreemfs.common.libxtreemfs.Options;
import org.xtreemfs.common.libxtreemfs.exceptions.PosixErrorException;
import org.xtreemfs.foundation.SSLOptions;
import org.xtreemfs.foundation.TimeSync;
import org.xtreemfs.foundation.VersionManagement;
import org.xtreemfs.foundation.logging.Logging;
import org.xtreemfs.foundation.pbrpc.Schemes;
import org.xtreemfs.foundation.pbrpc.generatedinterfaces.RPC.UserCredentials;
import org.xtreemfs.foundation.util.CLIParser;
import org.xtreemfs.foundation.util.CLIParser.CliOption;
import org.xtreemfs.foundation.util.OutputUtils;
import org.xtreemfs.pbrpc.generatedinterfaces.GlobalTypes.PORTS;
import org.xtreemfs.pbrpc.generatedinterfaces.GlobalTypes.SYSTEM_V_FCNTL;
import org.xtreemfs.pbrpc.generatedinterfaces.MRC.DirectoryEntries;
import org.xtreemfs.pbrpc.generatedinterfaces.MRC.DirectoryEntry;
import org.xtreemfs.pbrpc.generatedinterfaces.MRC.XATTR_FLAGS;
import org.xtreemfs.utils.DefaultDirConfig;
import org.xtreemfs.utils.utils;

/**
 * 
 * @author bjko
 */
public class xtfs_scrub {

    public static interface FileScrubbedListener {
        public void fileScrubbed(String FileName, long bytesScrubbed, Collection<ReturnStatus> rstatus);
    }

    public static enum ReturnStatus {
        FILE_OK, FILE_LOST, WRONG_FILE_SIZE, FAILURE_OBJECTS, FAILURE_REPLICAS, UNREACHABLE;
    };

    public static String                latestScrubAttr     = "scrubber.latestscrub";

    public static final UserCredentials credentials;

    static {
        credentials = UserCredentials.newBuilder().setUsername("root").addGroups("root").build();
    }

    private static final int            DEFAULT_NUM_THREADS = 10;

    private final AdminVolume                 volume;

    private final boolean               repair, delete, silent;

    private final ExecutorService       tPool;

    private long                        lastBytesScrubbed;

    private final Stack<String>         directories;

    private final Stack<String>         files;

    private final Object                completeLock;

    private int                         returnCode;

    private int                         numInFlight;

    private final int                   numThrs;

    private boolean                     hasFinished;

    private boolean                     isLatestScrubAttrSettable;

    private String                      currentDirName      = null;

    private int                         numFiles, numReplicaFailure, numObjectFailure, numFileOk,
            numUnreachable, numWrongFS, numDead;

    private final Set<String>           removedOSDs;

    public xtfs_scrub(AdminClient client, AdminVolume volume, int numThrs, boolean repair, boolean delete,
            boolean silent) throws IOException {
        this.repair = repair;
        this.delete = delete;
        this.silent = silent;
        this.numThrs = numThrs;
        directories = new Stack<String>();
        files = new Stack<String>();
        tPool = Executors.newFixedThreadPool(numThrs);
        numInFlight = 0;
        completeLock = new Object();
        hasFinished = false;
        isLatestScrubAttrSettable = true;

        this.volume = volume;

        numFiles = 0;
        numReplicaFailure = 0;
        numObjectFailure = 0;
        numFileOk = 0;
        numUnreachable = 0;

        if (!repair && !delete) {
            System.out.println("running in check mode, no changes to files will be made");
        } else {
            if (repair)
                System.out.println("running in repair mode");
            if (delete)
                System.out
                        .println("WARNING: delete is enabled, corrupt files (data lost due to failed OSDs) will be deleted");
        }

        removedOSDs = client.getRemovedOsds();

        if (removedOSDs.size() > 0) {
            System.out
                    .println("list of OSDs that have been removed (replicas on these OSDs will be deleted):");
            for (String uuid : removedOSDs) {
                System.out.println("\t" + uuid);
            }
        }

    }

    public int scrub() {

        System.out.println("");
        directories.push("/");
        // create scrub xattr if not done yet
        try {
            volume.setXAttr(credentials, "/", latestScrubAttr, Long.toString(TimeSync.getLocalSystemTime()),
                    XATTR_FLAGS.XATTR_FLAGS_CREATE);
        } catch (PosixErrorException e) {
            // scrub xattr already exists, nothing to do here.

        } catch (IOException e) {
            isLatestScrubAttrSettable = false;
            System.out.println("\nWarning: cannot mark volume as scrubbed: " + e);
        }
        fillQueue();
        synchronized (completeLock) {
            try {
                if (!hasFinished) {
                    completeLock.wait();
                }
            } catch (InterruptedException ex) {
            }
        }
        tPool.shutdown();
        try {
            tPool.awaitTermination(1, TimeUnit.HOURS);
        } catch (InterruptedException e) {
        }

        if (!silent)
            System.out.format("scrubbed %-42s      %15s - total %15s\n\u001b[100D\u001b[A", "all files", "",
                    OutputUtils.formatBytes(lastBytesScrubbed));

        System.out.println("\n\nsummary:");
        System.out.println("files checked                                                   : " + numFiles);
        System.out.println("  files ok                                                      : " + numFileOk);
        System.out.println("  files corrupted                                               : " + (numFiles - numFileOk));
        System.out.println("    of which had lost replicas (caused by removed OSDs)         : " + numReplicaFailure);
        System.out.println("    of which had corrupted objects (caused by invalid checksums): " + numObjectFailure);
        System.out.println("    of which had a wrong file size                              : " + numWrongFS);
        System.out.println("    of which are lost (unrecoverable)                           : " + numDead);
        System.out.println("    of which are unreachable (caused by communication errors)   : " + numUnreachable);
        System.out.println("bytes checked                                                   : " + OutputUtils.formatBytes(lastBytesScrubbed));

        return returnCode;
    }

    private void fillQueue() {
        try {
            synchronized (directories) {
                while (numInFlight < numThrs) {
                    try {
                        String fileName = files.pop();
                        try {

                            FileScrubbedListener fsListener = new FileScrubbedListener() {
                                @Override
                                public void fileScrubbed(String fileName, long bytesScrubbed,
                                        Collection<ReturnStatus> rstatus) {
                                    xtfs_scrub.this.fileScrubbed(fileName, bytesScrubbed, rstatus);
                                }
                            };

                            FileScrubber fi = new FileScrubber(fileName, volume, fsListener, removedOSDs,
                                    repair, delete);
                            tPool.submit(fi);
                            numInFlight++;
                        } catch (IOException ex) {
                            Logging.logError(Logging.LEVEL_WARN, this, ex);
                        }
                    } catch (EmptyStackException ex) {
                        // fetch next dir
                        fetchNextDir();
                    }
                }
            }
        } catch (EmptyStackException ex) {
            // no more entries, finished!
            if (isLatestScrubAttrSettable) {
                try {
                    // mark volume as scrubbed
                    volume.setXAttr(credentials, "/", latestScrubAttr,
                            Long.toString(TimeSync.getLocalSystemTime()), XATTR_FLAGS.XATTR_FLAGS_REPLACE);

                } catch (IOException ex2) {
                    System.out.println("\nWarning: cannot mark volume as successfully scrubbed: " + ex2);
                }
            }

            finish(0);
            return;
        }
    }

    private void finish(int returnCode) {
        synchronized (directories) {
            if (numInFlight == 0) {
                synchronized (completeLock) {
                    this.returnCode = returnCode;
                    this.hasFinished = true;
                    completeLock.notifyAll();
                }
            }
        }

    }

    public void fileScrubbed(String fileName, long bytesScrubbed, Collection<ReturnStatus> rs) {

        // update statistics

        if (fileName.length() > 42) {
            fileName = "..." + fileName.substring(fileName.length() - 39, fileName.length());
        }
        synchronized (directories) {
            lastBytesScrubbed += bytesScrubbed;
            numFiles++;
            for (ReturnStatus status : rs) {
                switch (status) {
                case FILE_OK:
                    numFileOk++;
                    break;
                case WRONG_FILE_SIZE:
                    numWrongFS++;
                    break;
                case UNREACHABLE:
                    numUnreachable++;
                    break;
                case FILE_LOST:
                    numDead++;
                    break;
                case FAILURE_REPLICAS:
                    numReplicaFailure++;
                    break;
                case FAILURE_OBJECTS:
                    numObjectFailure++;
                    break;
                }
            }
            if (!silent)
                System.out.format("scrubbed %-42s with %15s - total %15s\n\u001b[100D\u001b[A", fileName,
                        OutputUtils.formatBytes(bytesScrubbed), OutputUtils.formatBytes(lastBytesScrubbed));
            numInFlight--;
            fillQueue();
        }

    }

    private void fetchNextDir() {
        currentDirName = directories.pop();

        try {

            DirectoryEntries ls = volume.readDir(credentials, currentDirName, 0, 0, false);

            if (ls == null) {
                Logging.logMessage(Logging.LEVEL_ERROR, this, "path %s does not exist!", currentDirName);
                return;
            }

            for (int i = 0; i < ls.getEntriesCount(); i++) {
                DirectoryEntry e = ls.getEntries(i);
                if ((e.getStbuf().getMode() & SYSTEM_V_FCNTL.SYSTEM_V_FCNTL_H_S_IFREG.getNumber()) != 0) {
                    // regular file
                    files.push(currentDirName + e.getName());
                } else if ((e.getStbuf().getMode() & SYSTEM_V_FCNTL.SYSTEM_V_FCNTL_H_S_IFDIR.getNumber()) != 0) {
                    if (!e.getName().equals(".") && !e.getName().equals(".."))
                        directories.push(currentDirName + e.getName() + "/");
                }
            }
        } catch (IOException ex) {
            System.err.println("cannot contact MRC... aborting");
            System.err.println(ex);
        } catch (Exception ex) {
            ex.printStackTrace();
            throw new EmptyStackException();
        }

    }

    public static void main(String[] args) {

        Logging.start(Logging.LEVEL_WARN);

        System.out.println("XtreemFS scrubber version " + VersionManagement.RELEASE_VERSION
                + " (file system data integrity check)\n");

        Map<String, CliOption> options = utils.getDefaultAdminToolOptions(false);
        List<String> arguments = new ArrayList<String>(1);
        CliOption oDir = new CliOption(
                CliOption.OPTIONTYPE.STRING,
                "directory service to use (e.g. 'pbrpc://localhost:32638'). If no URI is specified, URI and security settings are taken from '"
                        + DefaultDirConfig.DEFAULT_DIR_CONFIG + "'", "<uri>");
        oDir.urlDefaultPort = PORTS.DIR_PBRPC_PORT_DEFAULT.getNumber();
        oDir.urlDefaultProtocol = Schemes.SCHEME_PBRPC;
        options.put("dir", oDir);
        options.put("repair", new CliOption(CliOption.OPTIONTYPE.SWITCH,
                "repair files (update file size, replace replicas) when possible", ""));
        options.put("delete", new CliOption(CliOption.OPTIONTYPE.SWITCH,
                "delete lost files (incomplete due to OSD failures)", ""));
        options.put("silent", new CliOption(CliOption.OPTIONTYPE.SWITCH, "don't show the progress bar", ""));
        options.put("thrs", new CliOption(CliOption.OPTIONTYPE.NUMBER,
                "number of concurrent file scrub threads (default=" + DEFAULT_NUM_THREADS + ")", "n"));

        CLIParser.parseCLI(args, options, arguments);

        if (arguments.size() != 1)
            error("invalid number of arguments", options);

        if (options.get(utils.OPTION_HELP).switchValue || options.get(utils.OPTION_HELP_LONG).switchValue) {
            usage(options);
            System.exit(0);
        }

        String[] dirURLs = (options.get("dir").stringValue != null) ? options.get("dir").stringValue
                .split(",") : null;

        SSLOptions sslOptions = null;
        String[] dirAddrs = null;

        if (dirURLs != null) {
            int i = 0;
            boolean gridSSL = false;
            dirAddrs = new String[dirURLs.length];
            for (String dirURL : dirURLs) {

                // parse security info if protocol is 'https'
                if (dirURL.contains(Schemes.SCHEME_PBRPCS + "://")
                        || dirURL.contains(Schemes.SCHEME_PBRPCG + "://") && sslOptions == null) {
                    String serviceCredsFile = options.get(utils.OPTION_USER_CREDS_FILE).stringValue;
                    String serviceCredsPass = options.get(utils.OPTION_USER_CREDS_PASS).stringValue;
                    if(serviceCredsPass != null && serviceCredsPass.equals("-")) {
                    	serviceCredsPass = new String(System.console().readPassword("Enter credentials password: "));
                    }
                    String trustedCAsFile = options.get(utils.OPTION_TRUSTSTORE_FILE).stringValue;
                    String trustedCAsPass = options.get(utils.OPTION_TRUSTSTORE_PASS).stringValue;
                    if(trustedCAsPass != null && trustedCAsPass.equals("-")) {
                    	trustedCAsPass = new String(System.console().readPassword("Enter trust store password: "));
                    }
                    String sslProtocolString = options.get(utils.OPTION_SSL_PROTOCOL).stringValue;
                    if (dirURL.contains(Schemes.SCHEME_PBRPCG + "://")) {
                        gridSSL = true;
                    }

                    if (serviceCredsFile == null) {
                        System.out.println("SSL requires '-" + utils.OPTION_USER_CREDS_FILE
                                + "' parameter to be specified");
                        usage(options);
                        System.exit(1);
                    } else if (trustedCAsFile == null) {
                        System.out.println("SSL requires '-" + utils.OPTION_TRUSTSTORE_FILE
                                + "' parameter to be specified");
                        usage(options);
                        System.exit(1);
                    }

                    // TODO: support custom SSL trust managers
                    try {
                        sslOptions = new SSLOptions(serviceCredsFile, serviceCredsPass, SSLOptions.PKCS12_CONTAINER,
                                trustedCAsFile, trustedCAsPass, SSLOptions.JKS_CONTAINER, false, gridSSL,
                                sslProtocolString, null);
                    } catch (Exception e) {
                        System.err.println("unable to set up SSL, because:" + e.getMessage());
                        System.exit(1);
                    }
                }

                // add URL to dirAddrs
                if (dirURL.contains("://")) {
                    // remove Protocol information
                    String[] tmp = dirURL.split("://");
                    // remove possible slash
                    dirAddrs[i++] = tmp[1].replace("/", "");
                } else {
                    // remove possible slash
                    dirAddrs[i++] = dirURL.replace("/", "");
                }
            }
        }

        // read default settings
        if (dirURLs == null) {
            try {
                DefaultDirConfig cfg = new DefaultDirConfig();
                sslOptions = cfg.getSSLOptions();
                dirAddrs = cfg.getDirectoryServices();
            } catch (Exception e) {
                System.err.println("unable to get SSL options, because: " + e.getMessage());
                System.exit(1);
            }
        }

        boolean repair = options.get("repair").switchValue;
        boolean silent = options.get("silent").switchValue;
        boolean delete = options.get("delete").switchValue;

        int numThreads = DEFAULT_NUM_THREADS;
        if (options.get("thrs").numValue != null) {
            numThreads = options.get("thrs").numValue.intValue();
        }

        final String volumeName = arguments.get(0);

        Options userOptions = new Options();

        // TODO: NativeClient necessary due to implemented quota features in C++ Client: finalize and clear voucher on
        // close
        AdminClient c = ClientFactory.createAdminClient(ClientType.NATIVE, dirAddrs, credentials, sslOptions,
                userOptions);
        AdminVolume volume = null;
        try {
            c.start();
            volume = c.openVolume(volumeName, sslOptions, new Options());
            volume.start();
        } catch (Exception e) {
            System.err.println("unable to scrub Volume, because: " + e.getMessage());
            System.exit(1);
        }

        int exitCode = 1;
        try {

            xtfs_scrub scrubber = new xtfs_scrub(c, volume, numThreads, repair, delete, silent);
            exitCode = scrubber.scrub();
            if (exitCode == 0) {
                System.out.println("\n\nsuccessfully scrubbed volume '" + volumeName + "'");
            } else {
                System.out.println("\n\nscrubbing volume '" + volumeName + "' FAILED!");
            }

            System.exit(exitCode);
        } catch (Exception e) {
            System.err.println("unable to scrub Volume, because: " + e.getMessage());
            System.exit(1);
        }
        c.shutdown();

    }

    private static void error(String message, Map<String, CliOption> options) {
        System.err.println(message);
        System.out.println();
        usage(options);
        System.exit(1);
    }

    private static void usage(Map<String, CliOption> options) {
        System.out.println("usage: xtfs_scrub [options] <volume_name>");
        System.out.println("<volume_name> the volume to scrub\n");
        System.out.println("options:");

        utils.printOptions(options);
    }
}
